### PR TITLE
[search] Do not require building to have a postcode if city/street has.

### PR DESCRIPTION
**Pull request by tatiana-yan**

### CODE REVIEW

File: search/features_layer_matcher.hpp [hunk 1]
```diff
@@ -173,10 +173,14 @@ class FeaturesLayerMatcher
     for (size_t i = 0; i < pois.size(); ++i)
     {
       m_context->ForEachFeature(
-          MercatorBounds::RectByCenterXYAndSizeInMeters(poiCenters[i].m_point, kBuildingRadiusMeters),
+          MercatorBounds::RectByCenterXYAndSizeInMeters(poiCenters[i].m_point,
+                                                        kBuildingRadiusMeters),
           [&](FeatureType & ft) {
-            if (m_postcodes && !m_postcodes->HasBit(ft.GetID().m_index))
+            if (m_postcodes && !(m_postcodes->HasBit(ft.GetID().m_index) ||
+                                 m_postcodes->HasBit(GetMatchingStreet(ft))))
```
Comment (mpimenov) [inline search/features_layer_matcher.hpp]:
```
            if (m_postcodes && !m_postcodes->HasBit(ft.GetID().m_index) &&
                                 !m_postcodes->HasBit(GetMatchingStreet(ft)))
```

File: search/features_layer_matcher.hpp [hunk 2]
```diff
@@ -294,21 +298,21 @@ class FeaturesLayerMatcher
...
       ++numFilterInvocations;
       if ((numFilterInvocations & 0xFF) == 0)
         BailIfCancelled(m_cancellable);
 
-      if (std::binary_search(buildings.begin(), buildings.end(), id))
+      if (std::binary_search(buildings.begin(), buildings.end(), houseId))
         return true;
 
-      if (m_postcodes && !m_postcodes->HasBit(id))
+      if (m_postcodes && !(m_postcodes->HasBit(houseId) || m_postcodes->HasBit(streetId)))
```
Comment (mpimenov) [inline search/features_layer_matcher.hpp]:
```
if (m_postcodes && !m_postcodes->HasBit(houseId)  && !m_postcodes->HasBit(streetId))
```
?

File: search/search_integration_tests/processor_test.cpp [hunk 3]
```diff
@@ -2207,5 +2207,73 @@ UNIT_CLASS_TEST(ProcessorTest, SynonymMisprintsTest)
     TEST(ResultsMatch("se street", rules), ());
   }
 }
+
+UNIT_CLASS_TEST(ProcessorTest, TestCityPostcodes)
```
Comment (mpimenov) [inline search/search_integration_tests/processor_test.cpp]: s/TestCityPostcodes/CityPostcodes/ ?

File: search/search_integration_tests/processor_test.cpp [hunk 4]
```diff
@@ -2207,5 +2207,73 @@ UNIT_CLASS_TEST(ProcessorTest, SynonymMisprintsTest)
...
+  auto countryId = BuildCountry(countryName, [&](TestMwmBuilder & builder) {
+    builder.Add(marckolsheim);
+    builder.Add(street);
+    builder.Add(building4);
+  });
+
+  SetViewport(m2::RectD(-1, -1, 1, 1));
+  {
+    Rules rules{ExactMatch(countryId, building4), ExactMatch(countryId, street)};
+    // Test we not require building to have postcode if city has.
```
Comment (mpimenov) [inline search/search_integration_tests/processor_test.cpp]: s/Test we/Test that we/
s/not/do not/
the building, the city

**Comment by tatiana-yan** [comment]
JTMAC

File: search/features_layer_matcher.hpp (hunk 1 above)
Comment (tatiana-yan) [reply search/features_layer_matcher.hpp]: done

File: search/features_layer_matcher.hpp (hunk 2 above)
Comment (tatiana-yan) [reply search/features_layer_matcher.hpp]: done

File: search/search_integration_tests/processor_test.cpp (hunk 3 above)
Comment (tatiana-yan) [reply search/search_integration_tests/processor_test.cpp]: done

File: search/search_integration_tests/processor_test.cpp (hunk 4 above)
Comment (tatiana-yan) [reply search/search_integration_tests/processor_test.cpp]: done

**Pull request merged**